pkg/controller/tas/topology_ungater.go

```diff
@@ -0,0 +1,300 @@
+}
+
+var _ reconcile.Reconciler = (*topologyUngater)(nil)
+
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;update;patch;delete
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads,verbs=get;list;watch
+// +kubebuilder:rbac:groups=kueue.x-k8s.io,resources=workloads/status,verbs=get
+
+func newTopologyUngater(c client.Client) *topologyUngater {
+	return &topologyUngater{
+		client: c,
+	}
+}
+
+func (r *topologyUngater) setupWithManager(mgr ctrl.Manager, cfg *configapi.Configuration) (string, error) {
+	podHandler := podHandler{}
+	return TASTopologyUngater, ctrl.NewControllerManagedBy(mgr).
+		Named(TASTopologyUngater).
+		For(&kueue.Workload{}).
+		Watches(&corev1.Pod{}, &podHandler).
+		WithOptions(controller.Options{NeedLeaderElection: ptr.To(false)}).
+		WithEventFilter(r).
+		Complete(core.WithLeadingManager(mgr, r, &kueue.ClusterQueue{}, cfg))
+}
+
+var _ handler.EventHandler = (*podHandler)(nil)
+
+type podHandler struct {
+}
+
+func (h *podHandler) Create(_ context.Context, e event.CreateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	pod, isPod := e.Object.(*corev1.Pod)
+	if !isPod {
+		return
+	}
+	h.queueReconcileForPod(pod, q)
+}
+
+func (h *podHandler) Update(ctx context.Context, e event.UpdateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	oldPod, isOldPod := e.ObjectOld.(*corev1.Pod)
+	newPod, isNewPod := e.ObjectNew.(*corev1.Pod)
+	if !isOldPod || !isNewPod {
+		return
+	}
+	h.queueReconcileForPod(oldPod, q)
+	h.queueReconcileForPod(newPod, q)
+}
+
+func (h *podHandler) Delete(_ context.Context, e event.DeleteEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	pod, isPod := e.Object.(*corev1.Pod)
+	if !isPod {
+		return
+	}
+	h.queueReconcileForPod(pod, q)
+}
+
+func (h *podHandler) queueReconcileForPod(pod *corev1.Pod, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	if pod == nil {
+		return
+	}
+	if !utilpod.HasGate(pod, kueuealpha.TopologySchedulingGate) {
+		return
+	}
+	if wlName, found := pod.Annotations[kueuealpha.WorkloadAnnotation]; found {
+		q.AddAfter(reconcile.Request{NamespacedName: types.NamespacedName{
+			Name:      wlName,
+			Namespace: pod.Namespace,
+		}}, ungateBatchPeriod)
+	}
+}
+
+func (h *podHandler) Generic(context.Context, event.GenericEvent, workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+}
+
+func (r *topologyUngater) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("workload", req.NamespacedName.Name)
+	log.V(2).Info("Reconcile Topology Ungater")
+
+	wl := &kueue.Workload{}
+	if err := r.client.Get(ctx, req.NamespacedName, wl); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return reconcile.Result{}, err
+		}
+		log.Info("workload not found")
+		return reconcile.Result{}, nil
+	}
+	if wl.Status.Admission == nil {
+		log.Info("workload is not admitted")
+		return reconcile.Result{}, nil
+	}
```

Comment:
I prefer to keep the duplicate checks to avoid panicing on race conditions:
- the predicates schedule reconcile by workload key , but then the workload is evicted
- the workload is recreated with the same key, but admitted without TAS

Reply:
As I commented above, after refactoring, these verifications seem to be needed.